geomesa-kafka/geomesa-kafka-09-datastore/src/main/scala/org/locationtech/geomesa/kafka09/LiveKafkaConsumerFeatureSource.scala

```diff
@@ -0,0 +1,276 @@
+            fireEvent(KafkaFeatureEvent.changed(this, update.feature))
+            featureCache.createOrUpdateFeature(update)
+          }
+        case del: Delete            =>
+          fireEvent(KafkaFeatureEvent.removed(this, featureCache.features(del.id).sf))
+          featureCache.removeFeature(del)
+        case clr: Clear             =>
+          fireEvent(KafkaFeatureEvent.cleared(this))
+          featureCache.clear()
+        case m                      => throw new IllegalArgumentException(s"Unknown message: $m")
+      }
+    }
+
+  // optimized for filter.include
+  override def getCountInternal(query: Query): Int = featureCache.size(query.getFilter)
+
+  override def getReaderForFilter(f: Filter): FR = featureCache.getReaderForFilter(f)
+
+  override def canEvent: Boolean = true
+
+  // Lazily fires events.
+  def fireEvent(event: => FeatureEvent) = {
+    if (contentState.hasListener) {
+      contentState.fireFeatureEvent(event)
+    }
+  }
+
+  override def close(): Unit = {
+    running.set(false)
+    client.shutdown()
+    es.shutdownNow()
+    ses.shutdownNow()
+  }
+}
+
+import KafkaFeatureEvent._
+
+class KafkaFeatureEvent(source: AnyRef,
+                        eventType: FeatureEvent.Type,
+                        bounds: ReferencedEnvelope,
+                        val feature: SimpleFeature)
+  extends FeatureEvent(source, eventType, bounds, buildId(feature.getID)) {
+}
+
+object KafkaFeatureEvent {
+  val ff = CommonFactoryFinder.getFilterFactory2
+
+  def buildId(id: String): Filter = {
+    val fid = new FeatureIdImpl(id)
+    val set = new util.HashSet[FeatureId]
+    set.add(fid)
+
+    ff.id(set)
+  }
+
+  def buildBounds(feature: SimpleFeature): ReferencedEnvelope = {
+    try {
+      val geom = feature.getDefaultGeometry.asInstanceOf[Point]
+      val lon = geom.getX
+      val lat = geom.getY
+
+      ReferencedEnvelope.create(new Envelope(lon, lon, lat, lat), CRS_EPSG_4326)
+    } catch {
+      case t: Throwable =>
+        KafkaConsumerFeatureSource.wholeWorldBounds
+    }
+  }
+
+  def changed(src: SimpleFeatureSource, feature: SimpleFeature): FeatureEvent =
+    new KafkaFeatureEvent(this,
+      Type.CHANGED,
+      KafkaFeatureEvent.buildBounds(feature),
+      feature)
+
+  def removed(src: SimpleFeatureSource, feature: SimpleFeature): FeatureEvent =
+    new FeatureEvent(this,
+      Type.REMOVED,
+      KafkaFeatureEvent.buildBounds(feature),
+      KafkaFeatureEvent.buildId(feature.getID))
+
+  def cleared(src: SimpleFeatureSource): FeatureEvent =
+    new FeatureEvent(this,
+      Type.REMOVED,
+      KafkaConsumerFeatureSource.wholeWorldBounds,
+      Filter.INCLUDE)
+}
+
+/** @param sft the [[SimpleFeatureType]]
+  * @param expirationPeriod the number of milliseconds after write to expire a feature or ``None`` to not
+  *                         expire
+  * @param ticker used to determine elapsed time for expiring entries
+  */
+class LiveFeatureCache(override val sft: SimpleFeatureType,
```

Comment:
This could be split out too, but will eventually end up in *geomesa-memory* (https://geomesa.atlassian.net/browse/GEOMESA-1410)